embedded-graphics/src/draw_target.rs

```diff
@@ -99,118 +95,149 @@ use crate::{
+///     framebuffer: [u16; 64 * 64],
 ///     iface: SPI1,
 /// }
 ///
-/// impl FastExampleDisplay {
-///     /// Draw a rectangle using hardware accelerated commands
-///     pub fn fast_rectangle(
-///         &self,
-///         rect: &Styled<Rectangle, PrimitiveStyle<Gray8>>,
-///     ) -> Result<(), CommError> {
-///         // Send rectangle drawing commands to the display
+/// impl ExampleDisplay {
+///     /// Send buffer to the display
+///     pub fn flush(&self) -> Result<(), ()> {
+///         self.iface.send_bytes(&self.framebuffer)
+///     }
+///
+///     /// Send commands to the display
+///     pub fn send_commands(&self, commands: &[u8]) -> Result<(), core::convert::Infallible> {
+///         // Send data marked as commands to the display.
 ///
 ///         Ok(())
 ///     }
 /// }
 ///
-/// impl DrawTarget<Gray8> for FastExampleDisplay {
-///     type Error = CommError;
+/// impl DrawTarget for ExampleDisplay {
+///     type Color = Rgb565;
+///     type Error = core::convert::Infallible;
 ///
-///     /// Draw a `pixel` that has a color defined as `Gray8`
-///     fn draw_pixel(&mut self, pixel: Pixel<Gray8>) -> Result<(), Self::Error> {
-///         let Pixel(coord, color) = pixel;
+///     fn size(&self) -> Size {
+///         Size::new(64, 64)
+///     }
 ///
-///         // Send commands directly to the display to set an individual pixel to the given color
+///     fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+///     where
+///         I: IntoIterator<Item = Pixel<Self::Color>> {
+///         for Pixel(coord, color) in pixels.into_iter() {
+///             if let Ok((x @ 0..=63, y @ 0..=63)) = coord.try_into() {
+///                 let index: u32 = x + y * 64;
+///                 self.framebuffer[index as usize] = RawU16::from(color).into_inner();
+///             }
+///         }
 ///
 ///         Ok(())
 ///     }
 ///
-///     fn size(&self) -> Size {
-///         Size::new(64, 64)
-///     }
-///
-///     /// Use the accelerated method when drawing rectangles
-///     ///
-///     /// This method overrides the default implementation. If `fast_rectangle()` fails, the error
-///     /// will be propagated through this method.
-///     fn draw_rectangle(
-///         &mut self,
-///         item: &Styled<Rectangle, PrimitiveStyle<Gray8>>,
-///     ) -> Result<(), Self::Error> {
-///         self.fast_rectangle(item)
+///     fn fill_solid(&mut self, area: &Rectangle, color: Self::Color) -> Result<(), Self::Error> {
+///         // Zero-sized rectangles have no bottom right corner and therefore should not be drawn
+///         let  bottom_right = if let Some(bottom_right) = area.bottom_right() {
+///             bottom_right
+///         } else {
+///             return Ok(())
+///         };
+///
+///         self.send_commands(&[
+///             // Draw rectangle command
+///             0x22,
+///             // Top left X coordinate
+///             area.top_left.x as u8,
```

Comment:
By using a simple cast here this won't have the correct behavior for coordinates outside the display area.

I just noticed that the drivers will also need to handle rectangles that are partially outside the display area, which makes this more complex then expected. Maybe we should add a clipping function for rectangles to make this easier to implement correctly:
```rust
 fn fill_solid(&mut self, area: &Rectangle, color: Self::Color) -> Result<(), Self::Error> {
    let display_area = Rectangle::new(Point::zero(), display.size());
    let actual_area = area.clip(&display_area);  //clip calculates the intersection of area and display_area
    ...
}
```

Reply:
> By using a simple cast here this won't have the correct behavior for coordinates outside the display area.

I'm struggling to come up with a clean-ish solution to show in this demo code. We could go for `var.max(0).min(63) as u8`, or something similar to the current `if let Ok((x @ 0..=63, y @ 0..=63)) = coord.try_into() { ... }`, but both feel quite messy.

Reply:
If you intersect the `area` rectangle with a rectangle that spans the visible display area the result is always in the valid range and you can just cast the coordinates, if the intersection is not `None`.

Reply:
Thanks for the hint. I've pushed a change to that example. Should be safe but still reasonably easy to understand.